src/Orleans.Runtime/Transactions/TransactionAgent.cs

```diff
@@ -274,11 +302,87 @@ public bool IsAborted(long transactionId)
 
         }
 
+        private async Task CommitTransactions(List<TransactionInfo> committingTransactions,
+            HashSet<long> outstandingCommits)
+        {
+            var stopWatch = Stopwatch.StartNew();
+            try
+            {
+                metrics.BatchCommitTransactionsRequestsCounter++;
+                var commitResponse = await this.tmService.CommitTransactions(committingTransactions, outstandingCommits);
+                metrics.BatchCommitTransactionsRequestLatencyCounter += stopWatch.Elapsed;
+                stopWatch.Stop();
+                var commitResults = commitResponse.CommitResult;
+
+                // reply to clients with the outcomes we received from the TM.
+                foreach (var completedId in commitResults.Keys)
+                {
+                    outstandingCommits.Remove(completedId);
+
+                    TaskCompletionSource<bool> completion;
+                    if (commitCompletions.TryRemove(completedId, out completion))
+                    {
+                        if (commitResults[completedId].Success)
+                        {
+                            TransactionsStatisticsGroup.OnTransactionCommitted();
+                            completion.SetResult(true);
+                        }
+                        else
+                        {
+                            if (commitResults[completedId].AbortingException != null)
+                            {
+                                TransactionsStatisticsGroup.OnTransactionAborted();
+                                completion.SetException(commitResults[completedId].AbortingException);
+                            }
+                            else
+                            {
+                                TransactionsStatisticsGroup.OnTransactionInDoubt();
+                                completion.SetException(new OrleansTransactionInDoubtException(completedId));
+                            }
+                        }
+                    }
+                }
+
+                // Refresh cached values using new values from TM.
+                this.ReadOnlyTransactionId = Math.Max(this.ReadOnlyTransactionId,
+                    commitResponse.ReadOnlyTransactionId);
+                this.abortLowerBound = Math.Max(this.abortLowerBound, commitResponse.AbortLowerBound);
+                logger.Debug(ErrorCode.Transactions_ReceivedTMResponse,
+                    "{0} transactions committed. readOnlyTransactionId {1}, abortLowerBound {2}",
+                    committingTransactions.Count, ReadOnlyTransactionId, abortLowerBound);
+            }
+            catch (Exception e)
+            {
+                metrics.BatchCommitTransactionsRequestLatencyCounter += stopWatch.Elapsed;
+                stopWatch.Stop();
```

Comment:
If we're trying to handle timing of commit call regardless of error, should probably wrap the this.tmService.CommitTransactions(..) call in a try finally, with the metrics done in the finally, and leave this section to error handling.  The way this is written we could double count the metrics if an error occurs after the CommitTransactions call.